src/gc/impl/conservative/gc.d

```diff
@@ -2374,12 +2374,14 @@ struct Gcx
 
         {
             // lock roots and ranges around suspending threads b/c they're not reentrant safe
+            ConservativeGC._isLocked = true;
             rangesLock.lock();
             rootsLock.lock();
             scope (exit)
             {
                 rangesLock.unlock();
                 rootsLock.unlock();
+                ConservativeGC._isLocked = false;
```

Comment:
From your description the failure seems to be that we're not unlocking those ranges on failure, so it's just broken cleanup on Error, but setting the reentrant flag is quite an ugly hackaround.
An explicit `try / catch (Error) / cleanup / rethrow` works.